enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,390 @@
+
+- Updating routes from default certificates to certificate referenced via
+  secrets and vice-versa.
+- Updating secret/certificate referenced in routes and verify serving
+  certificate has been updated.
+- Updating secret/certificate with incorrect information and verify route
+  is not admitted due to validation failure. (eg: mismatched public and private key, etc)
+
+### Graduation Criteria
+
+This feature will initially be released as Tech Preview only. The e2e tests
+in openshift/origin will only be added when graduating this feature to GA.
+
+#### Dev Preview -> Tech Preview
+
+N/A. This feature will go directly to Tech Preview.
+
+#### Tech Preview -> GA (Future work)
+
+The e2e tests as part of openshift/origin should be consistently passing.
+The router will need to undergo performance testing as part of OCP payload
+to ensure the memory implications of creating and maintains all the active watches
+is verified to be efficient.
+
+Update API godoc to document that manual intervention is required for using
+`.spec.tls.externalCertificate`.
+
+Update implementation details to cover internal working of secret monitor.
+
+##### Future work
+
+The ingress-to-route controller in the route-controller-manager will need to
+be updated to ensure that the created routes use `.spec.tls.externalCertificate`
+instead of `.spec.tls.certificate`. Additional tests will need to be added into
+o/origin for this scenario.
+
+Current implementation does not use `caCert` in the secret to populate
+`.spec.tls.caCertificate`, this can be added in the future.
+
+#### Removing a deprecated feature
+
+N/A.
+
+### Upgrade / Downgrade Strategy
+
+On downgrades, all routes specifying `.spec.tls.externalCertificate` will switch over to use the default certificates
+unless the route is manually edited and the `.spec.tls` is updated.
+
+Upgrade strategy not considered since this feature is going to be added as TechPreviewNoUpgrade.
+
+### Version Skew Strategy
+
+This feature will be added as TechPreviewNoUprade.
+
+### Operational Aspects of API Extensions
+
+Route validation in the API server will be modified to validate the following scenarios,
+
+- Check if secret/certificate referenced under `.spec.tls.externalCertificate` exists.
+- Check if secret/certificate referenced under `.spec.tls.externalCertificate` is of the correct type.
+- Check if router service account has permissions to read referenced secret.
+- Check if route has only one of the fields set,
+  - `.spec.tls.certificate` and `.spec.tls.key`
+  - `.spec.tls.externalCertificate`
+
+#### Failure Modes
+
+##### Referenced secret not present
+
+In scenarios where the secret has not been created by third-party solutions
+like cert-manager, the route would not be successfully created due to the
+dependency. This route will be rejected by the API server with an
+`FieldValueNotFound` error and will contain the reason as `referenced secret not present`.
```

Comment:
This detail is not added in the implementation details for the api server validation point.

Reply:
Good catch, this scenario is not valid anymore since we should be allowed to create a route when the secret does not yet exist but the permission to read is created. This validation will only be part of the router.